### ISSUE TITLE
Consider a new cost function: minimize the position gap between tvec and tvec_prior_?

### ISSUE THREAD

**Issue by hiakru** [colmap/colmap#896]
There are prior information from image, such as GPS x, y , z or ECEF x, y, z.  The information could be extracted in image tvec_prior_. 
Would we consider implemented a new cost function in cost_functions.h, which will minimize the gap between tvec_ and tvec_prior_  during mapper?
https://github.com/colmap/colmap/blob/9949d4e702f783fbecbe8c31477a8451c2de436a/src/base/image.h#L232

Currently there is only BA for re-projection error minimization during mapper. If the image position is also considered in global BA, the 3D model will be much robuster, otherwise sometimes the model will twisted.

Of course, this new feature could be just an option, could be configurable, default disable.

I could help pull request this new feature, if this is a necessary feature in colmap design~

Thanks,
Wesley

**Comment by hiakru** [comment]
Recently I did some experiments, use GPS in start and end of a tunnel, the result will be great!
Show some dense result, could help solve some semantic map problems.:
![frame000474](https://user-images.githubusercontent.com/33151098/84462184-eed67b80-aca0-11ea-8a7b-38004678b348.png)


![luntou-tunnel-back-test](https://user-images.githubusercontent.com/33151098/84462734-593beb80-aca2-11ea-917a-c3affcba1913.gif)

**Comment by yuancaimaiyi** [comment]
@hiakru awesome work. I think this feature could help many people.At least I benefit a lot。look forward to you push it

**Comment by csunking** [comment]
@hiakru  If you use these prior sensor data (imu, gps etc.),  does that could get converge result？

**Comment by hiakru** [comment]
> @hiakru If you use these prior sensor data (imu, gps etc.), does that could get converge result？
yes, literally the the gps should be trustful, which means it's somehow absolute position. I had tried meter accuracy gps, could help control the structure of map. 
moreover, you could try global bundle without GPS constraint after mapper finished. I call it post bundle adjustment.

**Comment by yuancaimaiyi** [comment]
@csunking  I use prior gps position and gps yaw ,and modify the colmap .use global bundlle adjustment,but I modify the global bundle adjusment, making the bundle adjustment only optimize the rotation(because I only have yaw ,abscent the pitch and roll ),and fixed the position(because the rtk postion is accurate ).As he @hiakru  said, this is called post-processing. my result also is very good .later ,I will push it to colmap,When reconstructing from a known pose, usig  the global bundle adjustment can  fix the position ,only optimize the rotation.

**Comment by yuancaimaiyi** [comment]
@hiakru Sorry to bother you ,you said " minimize the gap between tvec_ and tvec_prior_ during mapper", For example, if I have gps points at the start and end of the tunnel,  during the mapper that my estimated trajectory and my end point and the starting point model aligner(do a similar transformation, since the mapper estimates no scale), and then minimize the transformed coordinates (x,y) and the a priori gps coordinates (x,y) (since there are only two a priori values for the start and end points, you can only do the similar transformations of the plane but not be three-dimensional.Is that so?
 thanks

**Comment by csunking** [comment]
@yuancaimaiyi I agree with what you said. If @hiakru just only align the start and end GPS point, that seems meaningless thing. Because colmap has a module called 'Aligin' or something. But you has given an good idea of 'post bundle adjustment' operation which not like my origin idea of 'in the maper process to consirain the prior sensor data'.

**Comment by yuancaimaiyi** [comment]
I  still hope @hiakru  to open  coding, and I will open  my coding , so that we can learn from each other and help more people !!!!!!!!!!!!!!1@csunking  we are all the same major. I will git my coding as soon as possible.

**Comment by hiakru** [comment]
> @hiakru Sorry to bother you ,you said " minimize the gap between tvec_ and tvec_prior_ during mapper", For example, if I have gps points at the start and end of the tunnel, during the mapper that my estimated trajectory and my end point and the starting point model aligner(do a similar transformation, since the mapper estimates no scale), and then minimize the transformed coordinates (x,y) and the a priori gps coordinates (x,y) (since there are only two a priori values for the start and end points, you can only do the similar transformations of the plane but not be three-dimensional.Is that so?
> thanks

Tried several proposals, the settled down one is not like that,  actually the start/end images (with GPS) and middle tunnel images(without GPS) will reconstruct into only one model, no need do model aligner for several models. Thus, the GPS (X, Y, Z) 3-d data could help constraint the model structure during mapper GBA.

I suggest this is a normal function, not sure why not implemented in colmap. From original colmap author consideration, it might be no necessary, make colmap more scalable for diff sensors(IMU/GPS), and colmap is only try to figure out visual problems.

So guys, i'm not sure  actually~

**Comment by hiakru** [comment]
Onemore, as you add the real world gps information into the GBA, the model will back the the really world scale. So the mapper RegisterNextImage will do a real world localization, except the start period in mapper before GBA. Actually this is the truly localization used in real production, such as AR.

**Comment by mosjoker** [comment]
It seem like the option bundle_use_gcp(or bundle_use_gps)  ?
I used  stereo and gps information to reconstruct the map, which has the really world scale.But I found it's hard to apply gps information to multi-view geometry.  someone has good idea?

**Comment by csunking** [comment]
Maybe, your GPS time and video images frame time was not synchronous. Or,  you had not given a appropriate distance parameter in BA.

**Comment by SBCV** [comment]
> Currently there is only BA for re-projection error minimization during mapper. If the image position is also considered in global BA, the 3D model will be much robuster, otherwise sometimes the model will twisted.

I experienced these twisted effects with Colmap a lot, while trying to reconstruct driving scenarios (i.e. forward motions). I finally ended up using OpenMVG for these type of situations. Thus, such an option would be really helpful.

**Comment by yuancaimaiyi** [comment]
@SBCV .How did you use openmvg sfm stage with gps for restrictions (like gps at both ends of a tunnel), can you share your experience, I think a lot of people would like to know!

**Comment by SBCV** [comment]
Oh, sorry for the misunderstanding. I didn't use GPS information, but OpenMVG produced already clean, non-twisted reconstructions

**Comment by hiakru** [comment]
> Oh, sorry for the misunderstanding. I didn't use GPS information, but OpenMVG produced already clean, non-twisted reconstructions

Woo...  @SBCV  Could you give a brief introduce on how OpenMVG solved that problems, and the improvements against colmap?
Previously I had tried OpenMVG, but not very clear about the reconstruction detail.

**Comment by SBCV** [comment]
Frankly, I did not checkout algorithmic/implementation details. I've used both libraries over the past few years - and usually both compute very good results. 
Only in the specific case of forward motions (as in driving scenarios), I was not able to reliably compute reasonable reconstructions with Colmap - most of the time the reconstructions were somehow twisted and not usable. 
But I guess it will be difficult to determine what exactly causes the difference - there are so many parameters to adjust.
Therefore, I find it really impressive that you could fix it with this cost function!

**Comment by Tonsty** [comment]
hi, @hiakru ，your result of road 3d reconstruction is very cool! Could you tell what kind of device you use to capture the source images and show some source image examples?

> Recently I did some experiments, use GPS in start and end of a tunnel, the result was great!
> Show some dense result, could help solve some semantic map problems.:
> ![frame000474](https://user-images.githubusercontent.com/33151098/84462184-eed67b80-aca0-11ea-8a7b-38004678b348.png)
> 
> ![video](https://user-images.githubusercontent.com/33151098/84462734-593beb80-aca2-11ea-917a-c3affcba1913.gif)

**Comment by hiakru** [comment]
> hi, @hiakru ，your result of road 3d reconstruction is very cool! Could you tell what kind of device you use to capture the source images and show some source image examples?
> 
> > Recently I did some experiments, use GPS in start and end of a tunnel, the result was great!
> > Show some dense result, could help solve some semantic map problems.:
> > ![frame000474](https://user-images.githubusercontent.com/33151098/84462184-eed67b80-aca0-11ea-8a7b-38004678b348.png)
> > ![video](https://user-images.githubusercontent.com/33151098/84462734-593beb80-aca2-11ea-917a-c3affcba1913.gif)

Hi, @Tonsty I think many device could generate this result, such as iphone. High quality/resolution is not required.
this are few examples:
![1587363163 048659](https://user-images.githubusercontent.com/33151098/94094739-d4d93700-fe52-11ea-8432-226de321e184.jpg)
![1587363303 249389](https://user-images.githubusercontent.com/33151098/94094748-db67ae80-fe52-11ea-9096-7988e1e01a1a.jpg)

**Comment by hiakru** [comment]
@chengxiaolong1  and other interests
sorry for later reply, recently very busy with projects, I will fork a branch to implement this feature

**Comment by yuancaimaiyi** [comment]
@hiakru 
I have tried the openmvg ,I found that openmvg has already implemented the feature you mentioned. Openmvg can minimize the gap of tvec and prior GPS.Moreover, the incremental SFM of openmvg is better than colmap.
result as follows ：
colmap incremental SFM  (bending map)
![colmap](https://user-images.githubusercontent.com/28293910/94106056-31495000-fe6d-11ea-9050-c0923b67be3a.gif)

openmvg incremental SFM 
![111](https://user-images.githubusercontent.com/28293910/94106444-fa276e80-fe6d-11ea-8bdd-7e782456cd26.gif)


openmvg incremental SFM use gps prior( minimize the position gap between tvec and tvec_prior)
![openmvg_use_gps_prior](https://user-images.githubusercontent.com/28293910/94106252-a0bf3f80-fe6d-11ea-9f94-16f7207b0efb.gif)
